catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/SourcePoller.java

```diff
@@ -13,110 +13,605 @@
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.ServiceEvent;
+import org.osgi.framework.ServiceReference;
+import org.osgi.framework.hooks.service.EventListenerHook;
+import org.osgi.framework.hooks.service.ListenerHook;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 /**
- * The SourcePoller is the scheduler of the task to poll all configured sources at a fixed interval
- * to determine their availability. It is created by the CatalogFramework's blueprint.
+ * The {@link SourcePoller} checks the availability of all configured {@link Source}s provides a
+ * non-blocking solution to check whether a {@link Source} is available. The {@link SourcePoller}
+ * can be scheduled to execute at a configured rate, i.e., the polling interval.
+ *
+ * <p>This class maintains a list of all of the {@link Source}s and their last known availability.
+ * {@link Source}s are added to this list when they are created or destroyed, and {@link Source}
+ * availability is re-checked when the {@link Source} is modified. A cached map is maintained of all
+ * the {@link Source}s and their last known availability, i.e., {@link SourceStatus}.
  *
- * <p>An isAvailable() method is included in this class so that the caller, nominally the
- * CatalogFramework, can retrieve the cached availability of a specific source, or have it polled on
- * demand if there is no availability status cached.
+ * @see Source#isAvailable()
+ * @see Source#isAvailable(SourceMonitor)
+ * @see ddf.catalog.CatalogFramework#getSourceInfo(ddf.catalog.operation.SourceInfoRequest)
  */
-public class SourcePoller {
+public class SourcePoller implements EventListenerHook {
+
   private static final Logger LOGGER = LoggerFactory.getLogger(SourcePoller.class);
 
-  private static final int INITIAL_DELAY = 0;
+  private static final int EXECUTOR_SERVICE_TERMINATION_TIMEOUT = 5;
+
+  private static final TimeUnit EXECUTOR_SERVICE_TERMINATION_TIMEOUT_TIME_UNIT = TimeUnit.SECONDS;
+
+  private static final long DEFAULT_AVAILABILITY_CHECK_TIMEOUT = 1;
+
+  private static final TimeUnit AVAILABILITY_CHECK_TIMEOUT_TIME_UNIT = TimeUnit.MINUTES;
+
+  private long availabilityCheckTimeout = DEFAULT_AVAILABILITY_CHECK_TIMEOUT;
 
-  private int interval = 1;
+  private List<ConnectedSource> connectedSources = Collections.emptyList();
 
-  private ScheduledExecutorService scheduler;
+  private List<FederatedSource> federatedSources = Collections.emptyList();
 
-  private ScheduledFuture<?> handle;
+  private List<CatalogProvider> catalogProviders = Collections.emptyList();
 
-  private SourcePollerRunner runner;
+  private List<CatalogStore> catalogStores = Collections.emptyList();
+
+  private final ExecutorService sourceAvailabilityChecksThreadPool;
+
+  private final Cache<SourceKey, CachedSourceAvailability> cache;
+
+  @Nullable private ExecutorService pollingExecutorService;
+
+  /** Does not start polling */
+  public SourcePoller() {
+    final ThreadFactory threadFactory =
```

Comment:
I think we usually pass in executors/threads via blueprint. This will also make it easier to test.

Reply:
I refactored the `ExecutorService` to be a constructor argument in `SourcePollerRunner`, but I left the 2 `ExecutorService`s in `SourcePoller` because these thread pools do not need to be mocked in the unit tests. Also, the `SourcePoller`s are instantiated with factory methods, so I would also need to pass the 2 `ExecutorService`s as arguments in factory methods